eth/consensus/clique/snapshot_manager.py

```diff
@@ -0,0 +1,311 @@
+            try:
+                # We might have it saved on disk
+                return self.get_snapshot_from_db(block_hash)
+            except SnapshotNotFound:
+                pass
+
+            try:
+                # Otherwise, we can retrieve it on the fly
+                header = self._chain_db.get_block_header_by_hash(block_hash)
+            except HeaderNotFound:
+                raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+            else:
+                if header.block_number != block_number:
+                    raise SnapshotNotFound(
+                        f"Can not get snapshot for {block_hash} at {block_number}"
+                    )
+                else:
+                    return self._create_snapshot_from_checkpoint_header(header)
+
+        raise SnapshotNotFound(f"Can not get snapshot for {block_hash} at {block_number}")
+
+    def add_snapshot(self, mutable_snapshot: MutableSnapshot) -> Snapshot:
+        """
+        Retrieve a ``Snapshot`` from the given ``mutable_snapshot`` and add it to the local cache.
+        """
+        snapshot = mutable_snapshot.get_frozen_clone()
+        self._snapshots[snapshot.block_hash] = snapshot
+        return snapshot
+
+    def persist_snapshot(self, snapshot: Snapshot) -> None:
+        """
+        Persist the given snapshot to the database.
+        """
+        key = make_snapshot_lookup_key(snapshot.block_hash)
+        with self._chain_db.db.atomic_batch() as db:
+            db[key] = encode_snapshot(snapshot)
+
+    def get_snapshot_from_db(self, block_hash: Hash32) -> Snapshot:
+        """
+        Retrieve a snapshot from the database. Raise ``SnapshotNotFound`` if it does not exist.
+        """
+        key = make_snapshot_lookup_key(block_hash)
+        try:
+            encoded_key = self._chain_db.db[key]
+        except KeyError as e:
+            raise SnapshotNotFound(
+                f"Can not get on-disk snapshot for {block_hash}"
+            )
+        else:
+            return decode_snapshot(encoded_key)
+
+    def is_valid_vote(self,
+                      signers: Sequence[Address],
+                      subject: Address,
+                      action: VoteAction) -> bool:
+        """
+        Check if a vote is semantically valid, meaning it is either voting against an existing
+        member or voting in favor of an address that is not yet a member.
+        """
+        signer_exists = subject in signers
+        signer_is_kicked = signer_exists and action is VoteAction.KICK
+        signer_is_nominated = not signer_exists and action is VoteAction.NOMINATE
+        return signer_is_kicked or signer_is_nominated
+
+    def cast(self, snapshot: MutableSnapshot, subject: Address, action: VoteAction) -> bool:
+        """
+        Cast a vote on a ``MutableSnapshot``.
+        """
+        if not self.is_valid_vote(snapshot.signers, subject, action):
+            return False
```

Comment:
This pattern requires call sites to check return values which is an error prone mechanism.  Did you consider using exception raising instead to require call sites to handle both cases?

Reply:
I copied this from an existing implementation (geth/nethermind). I know it goes against our usual way of how to approach validation so I changed it into a `validate_for` method that raises an exception. 

That said (and I remember we have discussed this before but I don't remember your answer) I'm still not sold on this part:

>This pattern requires call sites to check return values which is an error prone mechanism.

With exception raising call sites need to wrap things in an `try` / `except` plus pick the right kind of exception to check for. I'm not sure if it really is less error prone.

In my pre-snake-charmer life, I've always tried to follow the advice to only use exceptions for *exceptional* cases. A failed validation is nothing exceptional for a method where the only purpose is validation. It either is valid or it isn't. Both things are the expected outcome in this scenario.

This also seems in line with the answers given here: https://softwareengineering.stackexchange.com/questions/330824/function-returning-true-false-vs-void-when-succeeding-and-throwing-an-exception

But I would like to learn or be reminded of your argument of exceptions being less error prone.

Notice, I'm not arguing to keep this method (I already changed it) or change our policy, I just want to be reminded of the arguments against using return values for validation.

Reply:
FWIW, @carver had a similar argument over [here](https://github.com/ethereum/trinity/pull/1237#discussion_r341405779).

Reply:
```python
def cast_vote(vote):
    if vote.is_valid():
        ...
        return True
    else:
        return False

def do_votes():
    # alice allowed to vote
    # bob isn't allowed to vote
    for vote in (alice, bob):
        cast_vote(alice)
```

Someone looking at `do_vote` wouldn't realize that they were not calling `cast_vote` correctly (since they aren't looking at the return value).  This code would not throw an error and would run as-if it was working even though it is broken.

```python
def cast_vote(vote):
    if vote.is_valid():
        ...
    else:
        raise ValidationError(...)
```

This model will blow up, making it really clear that the caller is not correctly using `cast_vote` and quickly showing the mis-use because an exception occurs.

The short argument is:

- When you forget to check a return value you end up with a silent failure.
- When you forget to handle an exception you get a loud exception.